### ISSUE TITLE
Radius r not automatically casted to integer when inserted in grid dictionary can cause memory bugs

### ISSUE THREAD

**Issue by Tortar** [JuliaDynamics/Agents.jl#750]
I noticed there shoud still be a (less severe) memory bug with keys in the `offsets_within_radius` dictionary: if a user called with a non-integer radius  (e.g. 1.1 first time 1.4 second time, etc...) any function which in turn calls the `offsets_within_radius` function in `gridgeneral.jl`, the unbounded cache problem (or anyway a waste of memory and time even if just two non-integer value non casted to integer are passed) could come back again.
I think there is a design problem here on how the radius is passed around but I think we can do the following thing to improve this:  ceil or floor the radius inside `offsets_within_radius` (and `offsets_within_radius_0`) instead of doing it from the outside and then use `Int` for radius keys in the underlying dict (I think this shouldn't break any code) and when calling `calculate_offsets`. One possibility to simplify a bit the code involved is to create a utility function which transforms the value of radius to the correct integer in relation to the metric used, it could be called `normalize_radius` or something similar.

Besides this, I'm still actually a bit confused about why we sometimes ceil and sometimes floor the value of radius, what is the reason of not going with just one way or the other with all metrics? In my opinion, if this is done just for supporting some better approximation of nearby ids for continuous space then it should be taken into account just there, and then just floor (I think it would be better than ceil) in all cases inside the internal grid

**Comment by Tortar** [comment]
I actually verified that now changing `offsets_within_radius` eliminating the floor and ceiling, like this:

```julia
function calculate_offsets(space::AbstractGridSpace{D}, r::Real) where {D}
    r0 = floor(Int, r)
    if space.metric == :euclidean
        hypercube = CartesianIndices((repeat([(-r0):r0], D)...,))
        # select subset which is in Hypersphere
        βs = [Tuple(β) for β ∈ hypercube if LinearAlgebra.norm(β.I) ≤ r]
    elseif space.metric == :manhattan
        hypercube = CartesianIndices((repeat([(-r0):r0], D)...,))
        βs = [Tuple(β) for β ∈ hypercube if sum(abs.(β.I)) ≤ r0]
    elseif space.metric == :chebyshev
        βs = vec([Tuple(a) for a in Iterators.product([(-r0):r0 for φ in 1:D]...)])
    else
        error("Unknown metric type")
    end
    length(βs) == 0 && push!(βs, ntuple(i -> 0, Val{D}())) # ensure 0 is there
    return βs::Vector{NTuple{D, Int}}
end
```

passes all tests.

**Comment by Datseris** [comment]
Yes, i agree with the changes.

For the floor ceiling: I am not sure... I think it depends on the metric.

**Comment by Tortar** [comment]
Solved in #751
